pkg/controller/controller.go

```diff
@@ -689,6 +689,15 @@ func (p *csiProvisioner) Provision(ctx context.Context, options controller.Provi
 		}
 	}
 
+	pvReadOnly := false
+	volCaps := req.GetVolumeCapabilities()
+	// if the request only has one accessmode and if its ROX, set readonly to true
+	if len(volCaps) == 1 {
+		if volCaps[0].GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_SINGLE_NODE_READER_ONLY || volCaps[0].GetAccessMode().GetMode() == csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY {
+			pvReadOnly = true
+		}
+	}
+
```

Comment:
I am not sure if presence of ROX accessMode means volume should be controller published as readonly.  To me, these they can be orthogonal concerns. cc @bswartz

Reply:
true.. but regardless of how controller publish the PV source could still set it RO .

Reply:
yes but interpreting ROX as readonly in PV source should be done with caution. It could make the PV unmountable in case journal needs to be replayed or something else.

Reply:
Yes I was going to argue the same. Volumes should not be controller-published as read-only unless the PV is marked read-only. There are too many good reasons for publishing them writable to the node and then having the node re-mount them as read-only to the pod.

Journal-replays are one use case. Pending expand operations are another.

I interpret PVs with the readOnly: true flag to be PERMANENTLY read-only, which you almost never want in practice. We might want to make it easier to cause that PV flag to get set to true in some cases, but I think the decision for ControllerPublishVolume should look at that flag only and not look at access modes.

Reply:
I agree the FS operations may need it RW while controller publish. But SP driver is free to do the operation or default to RW while controller publishing or node staging. Isnt it ? 

>I interpret PVs with the readOnly: true flag to be PERMANENTLY read-only, which you almost never want in practice. 

I could think of a scenario where user provision a new volume as "RO" with dataSource set to `PVC` or `Snapshot`. Isnt it a valid use case where user want a PV to be READONLY? 


@saad-ali @msau42 @jsafrane @xing-yang inviting your thoughts .

Reply:
I said _almost_ never, and you put your finger on one of the valid cases.

And I apologize for not reading the actual code when I last responded. I was just responding to the comment. I see now that this access mode check is deciding what to set for the PV's readOnly value, not the publish check. I think this is a relatively safe change to make.

The benefit of this change, as I see it:
- Making the PV read only means the nodes won't be able to write to the volume, which prevents data corruption in the case of a rogue node or a container break-out attack.

The downsides of this change are:
- Filesystem volumes created in this way won't be resizeable, ever.
- In the case that the snapshot from which the filesystem volume was cloned was taken with an unclean journal, the resulting volume will be unmountable and the user won't have any recourse.

Reply:
>The downsides of this change are:
    Filesystem volumes created in this way won't be resizeable, ever.

If the FS volume is created with RO mode, I dont see a reason to expand the volume at all :).. That said, its actually an `empty volume got provisioned` + `no data` can be written any way. so I dont see a reason to expand it. more or less, we can block the RO volume resize as an additional measure too.


 >In the case that the snapshot from which the filesystem volume was cloned was taken with an unclean journal, the resulting volume will be unmountable and the user won't have any recourse.
>

If the volume got provisioned from a datasource with RO mode,  I am wondering what could cause an unclean journal? if that was the case, SP driver could have declared failure in very first attempt of provisioning itself or I would rather put it as a clone/restore failure from storage backend.  Also, even in this case, SP driver still have a provision of respecting the RO flag in the PV or discarding it while staging or publishing it.

Reply:
> If the volume got provisioned from a datasource with RO mode, I am wondering what could cause an unclean journal? if that was the case, SP driver could have declared failure in very first attempt of provisioning itself or I would rather put it as a clone/restore failure from storage backend. Also, even in this case, SP driver still have a provision of respecting the RO flag in the PV or discarding it while staging or publishing it.

This is the sequence of events that could lead to trouble:

1. Create an writable volume, with an ext4 (or other journaling) filesystem
2. Pod attaches to the volume, and begins writing data
3. Node containing that pod crashes, loses power, or otherwise dies suddenly
4. Take a snapshot of the volume before it is attached to another pod
5. Create a new volume from the snapshot, and mark it ROX
6. Due to this PR, the new volume is marked readonly=true
6. Try to attach a pod to the new volume, and because the volume is readonly, it's controller-published as readonly
7. When the pod tries to come up, the mount fails because the kernel is unable to replay the journal on a readonly device

It's likely that the node crash doesn't even need to happen. An ordinary snapshot taken at the wrong moment could result in having a snapshot of a volume with a journal that needs replaying to successfully mount the volume.

To reproduce this, you'd need to be taking snapshot of a volume undergoing relatively heavy write load, and then trying to mount a readonly clone of that snapshot. Multiple attempts might be needed because it will depend on exactly when the snapshot occurs relative to the kernel's FS flushes to disk.

Reply:
> 8. the pod tries to come up, the mount fails because the kernel is unable to replay the journal on a readonly device

I'd like to question this part. I expect filesystems to be prepared for this case. Consider a normal device that degrades such that writes start to fail. The normal reaction of the kernel is to remount the filesystem as read-only. Being able to mount read-only after a system reboot to extract more data seems like a sensible feature to support - much better than refusing to mount completely.

Whether data and files are consistent is another question, but it's not different from protecting against a sudden power loss.

Reply:
First of all, I want to be clear that I'm not trying to block this issue. I think it's a good change, and I'd like to see it merge.

I was just trying to answer the question of what potential downsides this change could have, because it does introduce this new corner case with journaled filesystems. We shouldn't be surprised to see support cases from people with problems mounting readonly volumes that were cloned from in-use filesystems or snapshots taken of in-use filesystems, if journaling is in use.

There are some workarounds we can either recommend to users, or possibly even build into our CSI drivers to automatically overcome this class of problems. The most obvious workaround is to create a volume that's writable and use that instead. Or, create a writable volume, attach that to a pod that does nothing (to let the journal replay), then clone that volume after the pod has terminated cleanly (so the journal is empty).

The specific kernel error I see when I reproduce this problem with ext4 is:
[587843.214656] EXT4-fs (nbd2): INFO: recovery required on readonly filesystem
[587843.214657] EXT4-fs (nbd2): write access unavailable, cannot proceed (try mounting with noload)

Possibly CSI drivers could use the -o noload option with ext4 filesystems when mounting readonly to avoid the problem.

When using XFS, the kernel errors look like this:
[588571.823845] XFS (nbd1): Mounting V5 Filesystem
[588571.881098] XFS (nbd1): recovery required on read-only device.
[588571.881100] XFS (nbd1): write access unavailable, cannot proceed.
[588571.881101] XFS (nbd1): log mount/recovery failed: error -30
[588571.881123] XFS (nbd1): log mount failed

Some searching turns up the -o norecovery options for XFS which might also work.

In summary, I like this change but I'd urge that CSI driver authors consider special handling for mounting of volumes published read-only at the controller level. And if users do get stuck with unmountable volumes, be ready to suggest the workaround of using a writable volume.

Reply:
@bswartz @pohly thanks for the inputs.. First of all good discussion about the concerned area. 

>
Create an writable volume, with an ext4 (or other journaling) filesystem
Pod attaches to the volume, and begins writing data
Node containing that pod crashes, loses power, or otherwise dies suddenly
Take a snapshot of the volume before it is attached to another pod
....
>

Let me some thoughts here:  In general,  a storage vendor  does  seizing of `in flight i/o`s  and also put write `barriers` in place to make sure the data is persisted correctly while snapshot is taken. Thats one of the main requirement of consistent snapshots. Thus, in above scenario, it would be ideally achieved by the SP. From CSI layer, I assume it  should not be of much concern.  iow, a snapshot  taken on an `in used volume` or  snapshot taken after a crash or node/pod loss comes pretty much to the same situation and SP or any layer responsible at storage low level should take care of it.  While it comes to controller publish/node stage, again, decision can be made by SP to node stage or controller publish either by respecting the pv spec or not. 
Regardless, the mention about the possibilities of a 'unclean journal'  situation and a workaround/solution is indeed a good point :+1: . I am not sure about the different mount options of ext4 or xfs mentioned, have to look into the details of these. 
About the direction of this PR and its completion, I feel its good wait some more input from storage sig.

Reply:
Probably the best thing to do before merging this fix would be to add documentation to the CSI spec around the PUBLISH_READONLY controller capability. Drivers that assert that capability will need to be prepared to mount unmodifiable volumes on the nodes, which means understanding how to deal with unclean journals of each of the fstypes that your driver supports.

Reply:
While we wait for others to share their thoughts, let me rebase the PR and keep it updated with latest code.